File: ports/python/src/worldcut/requirements.py
from __future__ import annotations

import re
from collections.abc import Mapping
from typing import cast

from .canonicalization import canonical_json, sha256_digest
from .models import (
    AcquisitionAction,
    AcquisitionActionType,
    AcquisitionOption,
    CommonValidTimeRequirement,
    DependencyRequirement,
    FrozenJson,
    JsonValue,
    NormalizedTimestamp,
    Observation,
    Requirement,
    RequirementResult,
    ResourceIdentity,
    ValidityInterval,
    ValueEqualsRequirement,
    thaw_json,
)

_CANONICAL_ARRAY_INDEX = re.compile(r"^(0|[1-9][0-9]*)$")


def _resource_json(resource: ResourceIdentity) -> JsonValue:
    return {
        "provider": resource.provider,
        "account": resource.account,
        "kind": resource.kind,
        "key": resource.key,
    }


def _interval_json(interval: ValidityInterval) -> JsonValue:
    return {"from": interval.from_text, "until": interval.until_text}


def _action(
    action_type: AcquisitionActionType,
    observation: Observation | None,
    role: str,
    description: str,
    expected: JsonValue,
) -> AcquisitionAction:
    if observation is None:
        cost = 1
    elif action_type == "FETCH_REQUIRED_METADATA":
        cost = max(1, (observation.acquisition_cost + 3) // 4)
    else:
        cost = observation.acquisition_cost
    expected_digest = "none" if expected is None else sha256_digest(expected)[:12]
    return {
        "id": f"{action_type.lower()}:{role}:{expected_digest}",
        "type": action_type,
        "role": role,
        "cost": cost,
        "description": description,
        "expected": expected,
    }


def _option(
    requirement_id: str,
    suffix: str,
    description: str,
    actions: list[AcquisitionAction],
) -> AcquisitionOption:
    return {
        "id": f"{requirement_id}:{suffix}",
        "description": description,
        "actions": actions,
    }


def _missing_roles_result(
    requirement: Requirement, roles: list[str]
) -> RequirementResult:
    actions = [
        _action(
            "REFRESH_OBSERVATION",
            None,
            role,
            f"Acquire an observation for role {role}.",
            None,
        )
        for role in roles
    ]
    return {
        "requirementId": requirement.id,
        "requirementType": requirement.type,
        "required": requirement.required,
        "status": "UNKNOWN",
        "summary": (
            f"No observations are bound to required role(s): {', '.join(roles)}."
        ),
        "details": {"missingRoles": cast(list[JsonValue], roles)},
        "acquisitionOptions": [
            _option(
                requirement.id,
                "acquire-missing-roles",
                "Acquire every missing role required to evaluate this requirement.",
                actions,
            )
        ],
    }


def evaluate_dependency(
    requirement: DependencyRequirement,
    observations: Mapping[str, Observation],
) -> RequirementResult:
    missing_roles = [
        role
        for role in (requirement.dependent_role, requirement.target_role)
        if role not in observations
    ]
    if missing_roles:
        return _missing_roles_result(requirement, missing_roles)
    dependent = observations[requirement.dependent_role]
    target = observations[requirement.target_role]
    dependency = next(
        (
            candidate
            for candidate in dependent.witness.dependencies
            if candidate.name == requirement.dependency_name
        ),
        None,
    )
    if dependency is None:
        actions = [
            _action(
                "FETCH_REQUIRED_METADATA",
                dependent,
                dependent.role,
                f"Fetch dependency metadata for {dependent.role}.",
                {
                    "dependencyName": requirement.dependency_name,
                    "targetResource": _resource_json(target.resource),
                },
            )
        ]
        if target.witness.version is None:
            actions.append(
                _action(
                    "FETCH_REQUIRED_METADATA",
                    target,
                    target.role,
                    f"Fetch the resource version for {target.role}.",
                    None,
                )
            )
        return {
            "requirementId": requirement.id,
            "requirementType": requirement.type,
            "required": requirement.required,
            "status": "UNKNOWN",
            "summary": (
                f"{dependent.role} does not expose dependency "
                f"{requirement.dependency_name}."
            ),
            "details": {
                "dependentRole": dependent.role,
                "targetRole": target.role,
                "missingDependency": requirement.dependency_name,
            },
            "acquisitionOptions": [
                _option(
                    requirement.id,
                    "fetch-dependency-metadata",
                    "Fetch all metadata required to compare the dependency.",
                    actions,
                )
            ],
        }
    if dependency.resource != target.resource:
        return {
            "requirementId": requirement.id,
            "requirementType": requirement.type,
            "required": requirement.required,
            "status": "VIOLATED",
            "summary": (
                f"{dependent.role} is bound to a different resource than {target.role}."
            ),
            "details": {
                "dependentResource": _resource_json(dependency.resource),
                "targetResource": _resource_json(target.resource),
            },
            "acquisitionOptions": [
                _option(
                    requirement.id,
                    "acquire-compatible-resource",
                    (
                        "Acquire dependent evidence bound to the selected "
                        "target resource."
                    ),
                    [
                        _action(
                            "ACQUIRE_COMPATIBLE_EVIDENCE",
                            dependent,
                            dependent.role,
                            (
                                f"Acquire {dependent.role} evidence for the "
                                f"selected {target.role} resource."
                            ),
                            {"targetResource": _resource_json(target.resource)},
                        )
                    ],
                )
            ],
        }
    if dependency.version is None or target.witness.version is None:
        actions = []
        if dependency.version is None:
            actions.append(
                _action(
                    "FETCH_REQUIRED_METADATA",
                    dependent,
                    dependent.role,
                    f"Fetch the dependency version for {dependent.role}.",
                    {"dependencyName": requirement.dependency_name},
                )
            )
        if target.witness.version is None:
            actions.append(
                _action(
                    "FETCH_REQUIRED_METADATA",
                    target,
                    target.role,
                    f"Fetch the resource version for {target.role}.",
                    None,
                )
            )
        return {
            "requirementId": requirement.id,
            "requirementType": requirement.type,
            "required": requirement.required,
            "status": "UNKNOWN",
            "summary": (
                f"Version evidence is incomplete for {requirement.description}."
            ),
            "details": {
                "dependencyVersion": dependency.version,
                "targetVersion": target.witness.version,
            },
            "acquisitionOptions": [
                _option(
                    requirement.id,
                    "fetch-all-version-metadata",
                    "Fetch every missing version needed for this comparison.",
                    actions,
                )
            ],
        }
    dependency_version = dependency.version
    target_version = target.witness.version
    if dependency_version != target_version:
        return {
            "requirementId": requirement.id,
            "requirementType": requirement.type,
            "required": requirement.required,
            "status": "VIOLATED",
            "summary": (
                f"{requirement.description}: {dependency_version} does not equal "
                f"{target_version}."
            ),
            "details": {
                "dependentRole": dependent.role,
                "dependencyVersion": dependency_version,
                "targetRole": target.role,
                "targetVersion": target_version,
                "relation": dependency.relation,
            },
            "acquisitionOptions": [
                _option(
                    requirement.id,
                    "acquire-compatible-dependent",
                    (
                        "Acquire dependent evidence bound to the selected "
                        "target version."
                    ),
                    [
                        _action(
                            "ACQUIRE_COMPATIBLE_EVIDENCE",
                            dependent,
                            dependent.role,
                            (
                                f"Acquire {dependent.role} evidence bound to "
                                f"{target_version}."
                            ),
                            {
                                "targetRole": target.role,
                                "targetVersion": target_version,
                            },
                        )
                    ],
                ),
                _option(
                    requirement.id,
                    "refresh-target",
                    "Refresh the target before selecting compatible evidence.",
                    [
                        _action(
                            "REFRESH_OBSERVATION",
                            target,
                            target.role,
                            (
                                f"Refresh {target.role} before selecting "
                                "compatible evidence."
                            ),
                            {
                                "dependentRole": dependent.role,
                                "dependentVersion": dependency_version,
                            },
                        )
                    ],
                ),
            ],
        }
    return {
        "requirementId": requirement.id,
        "requirementType": requirement.type,
        "required": requirement.required,
        "status": "SATISFIED",
        "summary": (
            f"{requirement.description}: both roles are bound to {dependency_version}."
        ),
        "details": {
            "dependentRole": dependent.role,
            "targetRole": target.role,
            "version": dependency_version,
        },
        "acquisitionOptions": [],
    }


def evaluate_common_valid_time(
    requirement: CommonValidTimeRequirement,
    observations_by_role: Mapping[str, Observation],
) -> RequirementResult:
    missing_roles = [
        role for role in requirement.roles if role not in observations_by_role
    ]
    observations = [
        observations_by_role[role]
        for role in requirement.roles
        if role in observations_by_role
    ]
    missing_validity = [
        observation
        for observation in observations
        if observation.witness.validity is None
    ]
    prerequisite_actions = [
        _action(
            "REFRESH_OBSERVATION",
            None,
            role,
            f"Acquire an observation for role {role}.",
            None,
        )
        for role in missing_roles
    ]
    prerequisite_actions.extend(
        _action(
            "FETCH_REQUIRED_METADATA",
            observation,
            observation.role,
            f"Fetch validity metadata for {observation.role}.",
            {"within": _interval_json(requirement.within)},
        )
        for observation in missing_validity
    )

    latest_start = requirement.within.start
    earliest_end = requirement.within.end
    for observation in observations:
        validity = observation.witness.validity
        if validity is None:
            continue
        latest_start = max(latest_start, validity.start)
        if validity.end is not None:
            earliest_end = (
                validity.end
                if earliest_end is None
                else min(earliest_end, validity.end)
            )

    missing_validity_roles = [observation.role for observation in missing_validity]
    if earliest_end is not None and latest_start >= earliest_end:
        options = []
        for observation in observations:
            refresh = _action(
                "REFRESH_OBSERVATION",
                observation,
                observation.role,
                (f"Refresh {observation.role} to seek a compatible validity window."),
                {"within": _interval_json(requirement.within)},
            )
            options.append(
                _option(
                    requirement.id,
                    f"refresh-{observation.role}",
                    (
                        f"Refresh {observation.role} and acquire every other "
                        "missing prerequisite."
                    ),
                    [refresh, *prerequisite_actions],
                )
            )
        return {
            "requirementId": requirement.id,
            "requirementType": requirement.type,
            "required": requirement.required,
            "status": "VIOLATED",
            "summary": (
                f"{requirement.description}: the known validity intervals "
                "do not overlap."
            ),
            "details": {
                "roles": cast(list[JsonValue], list(requirement.roles)),
                "latestStart": _format_timestamp(latest_start),
                "earliestEnd": _format_timestamp(earliest_end),
                "missingRoles": cast(list[JsonValue], missing_roles),
                "missingValidityRoles": cast(list[JsonValue], missing_validity_roles),
            },
            "acquisitionOptions": options,
        }

    end_text = _format_timestamp(earliest_end) if earliest_end is not None else None
    if missing_roles or missing_validity:
        return {
            "requirementId": requirement.id,
            "requirementType": requirement.type,
            "required": requirement.required,
            "status": "UNKNOWN",
            "summary": f"{requirement.description}: validity evidence is incomplete.",
            "details": {
                "roles": cast(list[JsonValue], list(requirement.roles)),
                "missingRoles": cast(list[JsonValue], missing_roles),
                "missingValidityRoles": cast(list[JsonValue], missing_validity_roles),
                "possibleKnownWindow": {
                    "from": _format_timestamp(latest_start),
                    "until": end_text,
                },
            },
            "acquisitionOptions": [
                _option(
                    requirement.id,
                    "acquire-all-validity-prerequisites",
                    "Acquire every missing observation and validity witness.",
                    prerequisite_actions,
                )
            ],
        }
    return {
        "requirementId": requirement.id,
        "requirementType": requirement.type,
        "required": requirement.required,
        "status": "SATISFIED",
        "summary": f"{requirement.description}: a common valid time exists.",
        "details": {
            "roles": list(requirement.roles),
            "commonWindow": {
                "from": _format_timestamp(latest_start),
                "until": end_text,
            },
        },
        "acquisitionOptions": [],
    }


def _format_timestamp(value: NormalizedTimestamp) -> str:
    return value.text


def _value_at_path(value: FrozenJson, path: tuple[str, ...]) -> tuple[bool, FrozenJson]:
    current = value
    for segment in path:
        if isinstance(current, tuple):
            if _CANONICAL_ARRAY_INDEX.fullmatch(segment) is None:
                return False, None
            index = int(segment)
            if index >= len(current):
                return False, None
            current = current[index]
        elif isinstance(current, Mapping):
            if segment not in current:
                return False, None
            current = current[segment]
        else:
            return False, None
    return True, current


def evaluate_value_equals(
    requirement: ValueEqualsRequirement,
    observations: Mapping[str, Observation],
) -> RequirementResult:
    if requirement.role not in observations:
        return _missing_roles_result(requirement, [requirement.role])
    observation = observations[requirement.role]
    display_path = ".".join(requirement.path)
    found, frozen_actual = _value_at_path(observation.value, requirement.path)
    expected = thaw_json(requirement.expected)
    if not found:
        return {
            "requirementId": requirement.id,
            "requirementType": requirement.type,
            "required": requirement.required,
            "status": "UNKNOWN",
            "summary": (
                f"{requirement.description}: value path {display_path} is missing."
            ),
            "details": {
                "role": requirement.role,
                "path": list(requirement.path),
                "expected": expected,
            },
            "acquisitionOptions": [
                _option(
                    requirement.id,
                    "acquire-value",
                    "Acquire evidence containing the required value path.",
                    [
                        _action(
                            "ACQUIRE_COMPATIBLE_EVIDENCE",
                            observation,
                            observation.role,
                            (
                                f"Acquire {observation.role} evidence containing "
                                f"{display_path}."
                            ),
                            {
                                "path": list(requirement.path),
                                "expected": thaw_json(requirement.expected),
                            },
                        )
                    ],
                )
            ],
        }
    actual = thaw_json(frozen_actual)
    if canonical_json(actual) != canonical_json(expected):
        return {
            "requirementId": requirement.id,
            "requirementType": requirement.type,
            "required": requirement.required,
            "status": "VIOLATED",
            "summary": (
                f"{requirement.description}: observed value does not equal "
                "the required value."
            ),
            "details": {
                "role": requirement.role,
                "path": list(requirement.path),
                "expected": expected,
                "actual": actual,
            },
            "acquisitionOptions": [
                _option(
                    requirement.id,
                    "refresh-value",
                    "Refresh the observation before evaluating the value again.",
                    [
                        _action(
                            "REFRESH_OBSERVATION",
                            observation,
                            observation.role,
                            (
                                f"Refresh {observation.role} before evaluating "
                                f"{display_path}."
                            ),
                            {
                                "path": list(requirement.path),
                                "expected": thaw_json(requirement.expected),
                            },
                        )
                    ],
                )
            ],
        }
    return {
        "requirementId": requirement.id,
        "requirementType": requirement.type,
        "required": requirement.required,
        "status": "SATISFIED",
        "summary": (
            f"{requirement.description}: observed value matches the requirement."
        ),
        "details": {
            "role": requirement.role,
            "path": list(requirement.path),
            "expected": expected,
        },
        "acquisitionOptions": [],
    }

File: ports/go/requirements.go
package worldcut

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var canonicalArrayIndex = regexp.MustCompile(`^(0|[1-9][0-9]*)$`)

func resourceJSON(resource ResourceIdentity) map[string]any {
	return map[string]any{
		"provider": resource.Provider,
		"account":  resource.Account,
		"kind":     resource.Kind,
		"key":      resource.Key,
	}
}

func intervalJSON(interval *ValidityInterval) map[string]any {
	var until any
	if interval.Until != nil {
		until = *interval.Until
	}
	return map[string]any{"from": interval.From, "until": until}
}

func sameResource(left, right ResourceIdentity) bool {
	return left == right
}

func action(actionType string, observation *Observation, role, description string, expected any) (AcquisitionAction, error) {
	var cost int64 = 1
	if observation != nil {
		cost = observation.AcquisitionCost
		if actionType == "FETCH_REQUIRED_METADATA" {
			cost = (cost + 3) / 4
			if cost < 1 {
				cost = 1
			}
		}
	}
	expectedDigest := "none"
	if expected != nil {
		digest, err := SHA256Digest(expected)
		if err != nil {
			return AcquisitionAction{}, err
		}
		expectedDigest = digest[:12]
	}
	return AcquisitionAction{
		ID:          strings.ToLower(actionType) + ":" + role + ":" + expectedDigest,
		Type:        actionType,
		Role:        role,
		Cost:        cost,
		Description: description,
		Expected:    expected,
	}, nil
}

func option(requirementID, suffix, description string, actions []AcquisitionAction) AcquisitionOption {
	return AcquisitionOption{
		ID:          requirementID + ":" + suffix,
		Description: description,
		Actions:     actions,
	}
}

func missingRolesResult(requirement Requirement, roles []string) (RequirementResult, error) {
	actions := make([]AcquisitionAction, 0, len(roles))
	for _, role := range roles {
		candidate, err := action("REFRESH_OBSERVATION", nil, role, "Acquire an observation for role "+role+".", nil)
		if err != nil {
			return RequirementResult{}, err
		}
		actions = append(actions, candidate)
	}
	return RequirementResult{
		RequirementID:   requirement.ID,
		RequirementType: requirement.Type,
		Required:        requirement.isRequired(),
		Status:          "UNKNOWN",
		Summary:         "No observations are bound to required role(s): " + strings.Join(roles, ", ") + ".",
		Details:         map[string]any{"missingRoles": roles},
		AcquisitionOptions: []AcquisitionOption{
			option(requirement.ID, "acquire-missing-roles", "Acquire every missing role required to evaluate this requirement.", actions),
		},
	}, nil
}

func evaluateDependency(requirement Requirement, observations map[string]*Observation) (RequirementResult, error) {
	var missing []string
	for _, role := range []string{requirement.DependentRole, requirement.TargetRole} {
		if observations[role] == nil {
			missing = append(missing, role)
		}
	}
	if len(missing) != 0 {
		return missingRolesResult(requirement, missing)
	}
	dependent := observations[requirement.DependentRole]
	target := observations[requirement.TargetRole]
	var dependency *DependencyWitness
	for i := range dependent.Witness.Dependencies {
		if dependent.Witness.Dependencies[i].Name == requirement.DependencyName {
			dependency = &dependent.Witness.Dependencies[i]
			break
		}
	}
	if dependency == nil {
		first, err := action(
			"FETCH_REQUIRED_METADATA",
			dependent,
			dependent.Role,
			"Fetch dependency metadata for "+dependent.Role+".",
			map[string]any{
				"dependencyName": requirement.DependencyName,
				"targetResource": resourceJSON(target.Resource),
			},
		)
		if err != nil {
			return RequirementResult{}, err
		}
		actions := []AcquisitionAction{first}
		if target.Witness.Version == nil {
			second, err := action("FETCH_REQUIRED_METADATA", target, target.Role, "Fetch the resource version for "+target.Role+".", nil)
			if err != nil {
				return RequirementResult{}, err
			}
			actions = append(actions, second)
		}
		return RequirementResult{
			RequirementID:   requirement.ID,
			RequirementType: requirement.Type,
			Required:        requirement.isRequired(),
			Status:          "UNKNOWN",
			Summary:         dependent.Role + " does not expose dependency " + requirement.DependencyName + ".",
			Details: map[string]any{
				"dependentRole":     dependent.Role,
				"targetRole":        target.Role,
				"missingDependency": requirement.DependencyName,
			},
			AcquisitionOptions: []AcquisitionOption{
				option(requirement.ID, "fetch-dependency-metadata", "Fetch all metadata required to compare the dependency.", actions),
			},
		}, nil
	}
	if !sameResource(dependency.Resource, target.Resource) {
		candidate, err := action(
			"ACQUIRE_COMPATIBLE_EVIDENCE",
			dependent,
			dependent.Role,
			"Acquire "+dependent.Role+" evidence for the selected "+target.Role+" resource.",
			map[string]any{"targetResource": resourceJSON(target.Resource)},
		)
		if err != nil {
			return RequirementResult{}, err
		}
		return RequirementResult{
			RequirementID:      requirement.ID,
			RequirementType:    requirement.Type,
			Required:           requirement.isRequired(),
			Status:             "VIOLATED",
			Summary:            dependent.Role + " is bound to a different resource than " + target.Role + ".",
			Details:            map[string]any{"dependentResource": resourceJSON(dependency.Resource), "targetResource": resourceJSON(target.Resource)},
			AcquisitionOptions: []AcquisitionOption{option(requirement.ID, "acquire-compatible-resource", "Acquire dependent evidence bound to the selected target resource.", []AcquisitionAction{candidate})},
		}, nil
	}
	if dependency.Version == nil || target.Witness.Version == nil {
		actions := []AcquisitionAction{}
		if dependency.Version == nil {
			candidate, err := action(
				"FETCH_REQUIRED_METADATA",
				dependent,
				dependent.Role,
				"Fetch the dependency version for "+dependent.Role+".",
				map[string]any{"dependencyName": requirement.DependencyName},
			)
			if err != nil {
				return RequirementResult{}, err
			}
			actions = append(actions, candidate)
		}
		if target.Witness.Version == nil {
			candidate, err := action("FETCH_REQUIRED_METADATA", target, target.Role, "Fetch the resource version for "+target.Role+".", nil)
			if err != nil {
				return RequirementResult{}, err
			}
			actions = append(actions, candidate)
		}
		var dependencyVersion any
		if dependency.Version != nil {
			dependencyVersion = *dependency.Version
		}
		var targetVersion any
		if target.Witness.Version != nil {
			targetVersion = *target.Witness.Version
		}
		return RequirementResult{
			RequirementID:   requirement.ID,
			RequirementType: requirement.Type,
			Required:        requirement.isRequired(),
			Status:          "UNKNOWN",
			Summary:         "Version evidence is incomplete for " + requirement.Description + ".",
			Details:         map[string]any{"dependencyVersion": dependencyVersion, "targetVersion": targetVersion},
			AcquisitionOptions: []AcquisitionOption{
				option(requirement.ID, "fetch-all-version-metadata", "Fetch every missing version needed for this comparison.", actions),
			},
		}, nil
	}
	dependencyVersion := *dependency.Version
	targetVersion := *target.Witness.Version
	if dependencyVersion != targetVersion {
		dependentAction, err := action(
			"ACQUIRE_COMPATIBLE_EVIDENCE",
			dependent,
			dependent.Role,
			"Acquire "+dependent.Role+" evidence bound to "+targetVersion+".",
			map[string]any{"targetRole": target.Role, "targetVersion": targetVersion},
		)
		if err != nil {
			return RequirementResult{}, err
		}
		targetAction, err := action(
			"REFRESH_OBSERVATION",
			target,
			target.Role,
			"Refresh "+target.Role+" before selecting compatible evidence.",
			map[string]any{"dependentRole": dependent.Role, "dependentVersion": dependencyVersion},
		)
		if err != nil {
			return RequirementResult{}, err
		}
		return RequirementResult{
			RequirementID:   requirement.ID,
			RequirementType: requirement.Type,
			Required:        requirement.isRequired(),
			Status:          "VIOLATED",
			Summary:         fmt.Sprintf("%s: %s does not equal %s.", requirement.Description, dependencyVersion, targetVersion),
			Details: map[string]any{
				"dependentRole":     dependent.Role,
				"dependencyVersion": dependencyVersion,
				"targetRole":        target.Role,
				"targetVersion":     targetVersion,
				"relation":          dependency.Relation,
			},
			AcquisitionOptions: []AcquisitionOption{
				option(requirement.ID, "acquire-compatible-dependent", "Acquire dependent evidence bound to the selected target version.", []AcquisitionAction{dependentAction}),
				option(requirement.ID, "refresh-target", "Refresh the target before selecting compatible evidence.", []AcquisitionAction{targetAction}),
			},
		}, nil
	}
	return RequirementResult{
		RequirementID:      requirement.ID,
		RequirementType:    requirement.Type,
		Required:           requirement.isRequired(),
		Status:             "SATISFIED",
		Summary:            requirement.Description + ": both roles are bound to " + dependencyVersion + ".",
		Details:            map[string]any{"dependentRole": dependent.Role, "targetRole": target.Role, "version": dependencyVersion},
		AcquisitionOptions: []AcquisitionOption{},
	}, nil
}

func intervalTimes(interval *ValidityInterval) (time.Time, *time.Time) {
	start, _ := time.Parse(timestampLayout, interval.From)
	if interval.Until == nil {
		return start, nil
	}
	end, _ := time.Parse(timestampLayout, *interval.Until)
	return start, &end
}

func formatTime(value time.Time) string {
	return value.UTC().Format(timestampLayout)
}

func evaluateCommonValidTime(requirement Requirement, observationsByRole map[string]*Observation) (RequirementResult, error) {
	missingRoles := []string{}
	observations := []*Observation{}
	missingValidity := []*Observation{}
	for _, role := range requirement.Roles {
		observation := observationsByRole[role]
		if observation == nil {
			missingRoles = append(missingRoles, role)
			continue
		}
		observations = append(observations, observation)
		if observation.Witness.Validity == nil {
			missingValidity = append(missingValidity, observation)
		}
	}
	prerequisites := []AcquisitionAction{}
	for _, role := range missingRoles {
		candidate, err := action("REFRESH_OBSERVATION", nil, role, "Acquire an observation for role "+role+".", nil)
		if err != nil {
			return RequirementResult{}, err
		}
		prerequisites = append(prerequisites, candidate)
	}
	for _, observation := range missingValidity {
		candidate, err := action(
			"FETCH_REQUIRED_METADATA",
			observation,
			observation.Role,
			"Fetch validity metadata for "+observation.Role+".",
			map[string]any{"within": intervalJSON(requirement.Within)},
		)
		if err != nil {
			return RequirementResult{}, err
		}
		prerequisites = append(prerequisites, candidate)
	}
	latestStart, earliestEnd := intervalTimes(requirement.Within)
	for _, observation := range observations {
		if observation.Witness.Validity == nil {
			continue
		}
		start, end := intervalTimes(observation.Witness.Validity)
		if start.After(latestStart) {
			latestStart = start
		}
		if end != nil && (earliestEnd == nil || end.Before(*earliestEnd)) {
			copy := *end
			earliestEnd = &copy
		}
	}
	missingValidityRoles := make([]string, 0, len(missingValidity))
	for _, observation := range missingValidity {
		missingValidityRoles = append(missingValidityRoles, observation.Role)
	}
	if earliestEnd != nil && !latestStart.Before(*earliestEnd) {
		options := make([]AcquisitionOption, 0, len(observations))
		for _, observation := range observations {
			refresh, err := action(
				"REFRESH_OBSERVATION",
				observation,
				observation.Role,
				"Refresh "+observation.Role+" to seek a compatible validity window.",
				map[string]any{"within": intervalJSON(requirement.Within)},
			)
			if err != nil {
				return RequirementResult{}, err
			}
			actions := append([]AcquisitionAction{refresh}, prerequisites...)
			options = append(options, option(
				requirement.ID,
				"refresh-"+observation.Role,
				"Refresh "+observation.Role+" and acquire every other missing prerequisite.",
				actions,
			))
		}
		return RequirementResult{
			RequirementID:   requirement.ID,
			RequirementType: requirement.Type,
			Required:        requirement.isRequired(),
			Status:          "VIOLATED",
			Summary:         requirement.Description + ": the known validity intervals do not overlap.",
			Details: map[string]any{
				"roles":                requirement.Roles,
				"latestStart":          formatTime(latestStart),
				"earliestEnd":          formatTime(*earliestEnd),
				"missingRoles":         missingRoles,
				"missingValidityRoles": missingValidityRoles,
			},
			AcquisitionOptions: options,
		}, nil
	}
	var endJSON any
	if earliestEnd != nil {
		endJSON = formatTime(*earliestEnd)
	}
	if len(missingRoles) != 0 || len(missingValidity) != 0 {
		return RequirementResult{
			RequirementID:   requirement.ID,
			RequirementType: requirement.Type,
			Required:        requirement.isRequired(),
			Status:          "UNKNOWN",
			Summary:         requirement.Description + ": validity evidence is incomplete.",
			Details: map[string]any{
				"roles":                requirement.Roles,
				"missingRoles":         missingRoles,
				"missingValidityRoles": missingValidityRoles,
				"possibleKnownWindow":  map[string]any{"from": formatTime(latestStart), "until": endJSON},
			},
			AcquisitionOptions: []AcquisitionOption{
				option(requirement.ID, "acquire-all-validity-prerequisites", "Acquire every missing observation and validity witness.", prerequisites),
			},
		}, nil
	}
	return RequirementResult{
		RequirementID:   requirement.ID,
		RequirementType: requirement.Type,
		Required:        requirement.isRequired(),
		Status:          "SATISFIED",
		Summary:         requirement.Description + ": a common valid time exists.",
		Details: map[string]any{
			"roles":        requirement.Roles,
			"commonWindow": map[string]any{"from": formatTime(latestStart), "until": endJSON},
		},
		AcquisitionOptions: []AcquisitionOption{},
	}, nil
}

func valueAtPath(value any, path []string) (any, bool) {
	current := value
	for _, segment := range path {
		switch typed := current.(type) {
		case []any:
			if !canonicalArrayIndex.MatchString(segment) {
				return nil, false
			}
			index, err := strconv.ParseUint(segment, 10, 64)
			if err != nil || index >= uint64(len(typed)) {
				return nil, false
			}
			current = typed[index]
		case map[string]any:
			next, ok := typed[segment]
			if !ok {
				return nil, false
			}
			current = next
		default:
			return nil, false
		}
	}
	return current, true
}

func evaluateValueEquals(requirement Requirement, observations map[string]*Observation) (RequirementResult, error) {
	observation := observations[requirement.Role]
	if observation == nil {
		return missingRolesResult(requirement, []string{requirement.Role})
	}
	displayPath := strings.Join(requirement.Path, ".")
	actual, found := valueAtPath(observation.Value, requirement.Path)
	if !found {
		candidate, err := action(
			"ACQUIRE_COMPATIBLE_EVIDENCE",
			observation,
			observation.Role,
			"Acquire "+observation.Role+" evidence containing "+displayPath+".",
			map[string]any{"path": requirement.Path, "expected": requirement.Expected},
		)
		if err != nil {
			return RequirementResult{}, err
		}
		return RequirementResult{
			RequirementID:   requirement.ID,
			RequirementType: requirement.Type,
			Required:        requirement.isRequired(),
			Status:          "UNKNOWN",
			Summary:         requirement.Description + ": value path " + displayPath + " is missing.",
			Details:         map[string]any{"role": requirement.Role, "path": requirement.Path, "expected": requirement.Expected},
			AcquisitionOptions: []AcquisitionOption{
				option(requirement.ID, "acquire-value", "Acquire evidence containing the required value path.", []AcquisitionAction{candidate}),
			},
		}, nil
	}
	actualCanonical, err := CanonicalJSON(actual)
	if err != nil {
		return RequirementResult{}, err
	}
	expectedCanonical, err := CanonicalJSON(requirement.Expected)
	if err != nil {
		return RequirementResult{}, err
	}
	if string(actualCanonical) != string(expectedCanonical) {
		candidate, err := action(
			"REFRESH_OBSERVATION",
			observation,
			observation.Role,
			"Refresh "+observation.Role+" before evaluating "+displayPath+".",
			map[string]any{"path": requirement.Path, "expected": requirement.Expected},
		)
		if err != nil {
			return RequirementResult{}, err
		}
		return RequirementResult{
			RequirementID:   requirement.ID,
			RequirementType: requirement.Type,
			Required:        requirement.isRequired(),
			Status:          "VIOLATED",
			Summary:         requirement.Description + ": observed value does not equal the required value.",
			Details:         map[string]any{"role": requirement.Role, "path": requirement.Path, "expected": requirement.Expected, "actual": actual},
			AcquisitionOptions: []AcquisitionOption{
				option(requirement.ID, "refresh-value", "Refresh the observation before evaluating the value again.", []AcquisitionAction{candidate}),
			},
		}, nil
	}
	return RequirementResult{
		RequirementID:      requirement.ID,
		RequirementType:    requirement.Type,
		Required:           requirement.isRequired(),
		Status:             "SATISFIED",
		Summary:            requirement.Description + ": observed value matches the requirement.",
		Details:            map[string]any{"role": requirement.Role, "path": requirement.Path, "expected": requirement.Expected},
		AcquisitionOptions: []AcquisitionOption{},
	}, nil
}
